### ISSUE TITLE
Enable pre-set of username/password for GUI via env var, like STGUIAPIKEY

### ISSUE THREAD

**Issue by bluepuma77** [syncthing/syncthing#8791]
A lot of services are today deployed with containers on multiple machines. When looking at various containers like tools and databases, they all provide an option to set username/password up front via environment variables to make deployments easier. 

Examples: PHPAdmin, MongoExpress, MySQL, MongoDB

It would be great to have 2 environment variables to set username and password for the GUI, so it can be easily deployed via container with ports open and security already up in place.

Syncthing already support [STGUIAPIKEY](https://docs.syncthing.net/users/syncthing.html#development-settings) to set the API key via env var, which can be used with container deployments. It would be great to expand that to user/pass to easily enable usage of the GUI in a secured way.

**Comment by acolomb** [comment]
Is it really just username and password?  We support e.g. LDAP authentication as well, in the future possible WebAuthn.  So many options to configure overall, we'd need a whole bunch variables.  If the e.g. config file says LDAP and you set the env var STGUIUSERNAME, what authentication will be used?  Would this maybe open up some potential security issues?

**Comment by bluepuma77** [comment]
Well, I am not creating a configuration, I just run with the Syncthing container defaults. Just want to add an initial username/password for a bit improved security, like many other containers are providing.

I am currently creating a headless Syncthing controller for Docker Swarm, the Synching servers are just spun up with defaults. `STGUIAPIKEY` is there, the only thing missing is user/pass to be able to securely expose the GUI and have a look if everything is alright.

**Comment by jeremysprofile** [comment]
So this used to exist via `STGUIAUTH` (https://github.com/syncthing/syncthing/pull/557), got moved to a `--gui-authentication` flag (https://github.com/syncthing/syncthing/commit/214f18cbfdf7a3460c87892ad9b) and then both were removed (https://github.com/syncthing/syncthing/pull/2336) by @calmh because "this seems fundamentally insecure and I'm unsure of the actual use case for it".

Having the ability to turn on GUI authentication immediately with a preset password seems more secure than allowing immediate access. My setup is running Syncthing in Docker (using the host network) on my NAS. I can access the GUI via https://192.168.0.3:8384 from my laptop on the same local network, without authentication.

**Comment by acolomb** [comment]
I think the correct way here would be for the container to accept environment variables which are only used during initial creation of the config, maybe through some first-run setup script. That should cover your use case, right? Syncthing itself doesn't need to accept override variables for authentication once a config file exists which holds the once-configured values.

**Comment by jeremysprofile** [comment]
Yes, I would not want these env vars to override an existing config file. My use case is probably not worth talking about - I think I will just end up mucking with the config file and that will be good enough for me.

**Comment by dragetd** [comment]
I was just setting up syncthing in a container and came across the same issue.
This aligns with the concepts of a 12-factor-app https://12factor.net/config

Thea idea is that the configuration is separated from the runtime data and that it can be passed in externally. Some applications considered more 'cloud native' even enable setting ANY property through the config OR an environment variable.

The idea is: I might be happy with the default values for most settings and just want change a few things. These few things I would configure via environment variables. I would not even mount/persist the config at all!

Currently, the syncthing config contains the shared directories and a version number, if I saw this right. The 'shared directories' is more of a runtime-config and would have to move out of the config file. And there is also the index.

One way to approach this could be an entrypoint script that just collects all `ST_XXX_YYY`  variables and extracts XXX and YYY etc. and traverses the config XML, replacing each value with the value of the environment variable.

This would be a docker-only solution, but get us 98% to the goal of easier deploying syncthing in a cloud-native way.

A cleaner and more flexible approach, but also much more work to implement, would be to have a config system that loads config values with priority. Defaulting to built-in defaults, then checking a config file (which would not have to be complete) and then checking environment variables.

Then anyone would benefit from this and could configure things via environment variables.

**Comment by funkyfuture** [comment]
for the container image case; the `entrypoint.sh` could test whether user credentials exists in the config file and invoke `syncthing generate …` with a username and password/hash from environment variables. that'd require an xml parser and query tool available though :-/.

**Comment by xmha97** [comment]
If implemented in the following way, it would be more practical for everyone:

```yaml
name: homeserver
services:
  syncthing:
    image: syncthing/syncthing:latest
    container_name: my-syncthing
    hostname: hp-syncthing
    environment:
      - SYNCTHING_USER: ${SYNCTHING_USER} # GUI Authentication User
      - SYNCTHING_PASSWORD: ${SYNCTHING_PASSWORD} # GUI Authentication Password
      - SYNCTHING_DEVICE: ${SYNCTHING_DEVICE} # Device Name
```
